src/nnet3bin/nnet3-egs-augment-image.cc

```diff
@@ -147,37 +132,91 @@ void PerturbImage(const ImageAugmentationConfig &config,
                   MatrixBase<BaseFloat> *image) {
   config.Check();
   int32 image_width = image->NumRows(),
-      num_channels = config.num_channels,
-      image_height = image->NumCols() / num_channels;
+    num_channels = config.num_channels,
+    image_height = image->NumCols() / num_channels;
   if (image->NumCols() % num_channels != 0) {
     KALDI_ERR << "Number of columns in image must divide the number "
         "of channels";
   }
-  if (WithProb(config.horizontal_flip_prob)) {
-    HorizontalFlip(image);
-  }
-  { // horizontal shift
-    int32 horizontal_shift_max =
-        static_cast<int32>(0.5 + config.horizontal_shift * image_width);
-    if (horizontal_shift_max > image_width - 1)
-      horizontal_shift_max = image_width - 1;  // would be very strange.
-    int32 horizontal_shift = RandInt(-horizontal_shift_max,
-                                     horizontal_shift_max);
-    if (horizontal_shift != 0)
-      HorizontalShift(horizontal_shift, image);
-  }
-
-  { // vertical shift
-    int32 vertical_shift_max =
-        static_cast<int32>(0.5 + config.vertical_shift * image_height);
-    if (vertical_shift_max > image_height - 1)
-      vertical_shift_max = image_height - 1;  // would be very strange.
-    int32 vertical_shift = RandInt(-vertical_shift_max,
-                                     vertical_shift_max);
-    if (vertical_shift != 0)
-      VerticalShift(vertical_shift, num_channels, image);
-  }
 
+  // We do an affine transform which
+  // handles flipping, translation, rotation, magnification, and shear.
+
+  // Prepare the affine transform for geometric transformation
+  Matrix<BaseFloat> transform_mat(3, 3, kUndefined);
+  transform_mat.SetUnit();
+
+  Matrix<BaseFloat> shift_mat(3, 3, kUndefined);
+  shift_mat.SetUnit();
+  // translation (shift) mat:
+  // [ 1   0  x_shift
+  //   0   1  y_shift
+  //   0   0  1       ]
+
+  BaseFloat horizontal_shift = (2.0 * RandUniform() - 1.0) *
+                               config.horizontal_shift * image_width;
+  BaseFloat vertical_shift = (2.0 * RandUniform() - 1.0) *
+                             config.vertical_shift * image_height;
+  shift_mat(0, 2) = floor(horizontal_shift);  // best to floor shifts to avoid blurring
+  shift_mat(1, 2) = floor(vertical_shift);
+  // since we will center the image before applying the transform,
+  // horizontal flipping is simply achieved by setting [0, 0] to -1:
+  if (WithProb(config.horizontal_flip_prob))
+    shift_mat(0, 0) = -1;
```

Comment:
There might be some reason, or maybe not because I have tested this (alone and with other operations). I have also tested rotation. All works fine.